auto_sidecar.go

```diff
@@ -579,30 +637,67 @@ func (a *SidecarNegotiator) autoSidecarProvider(ctx context.Context,
 					break stateUpdateLoop
 				case newPktState.CurrentState == sidecar.StateExpectingChannel:
 					break stateUpdateLoop
+				case newPktState.CurrentState == sidecar.StateCanceled:
+					break stateUpdateLoop
 				}
 			}
 
-		case <-a.ticketFinalized:
-			log.Infof("Receiver negotiation for SidecarTicket(%x) "+
-				"complete!", localTicket.ID[:])
+		case fin := <-a.ticketFinalized:
+			log.Infof("Provider negotiation for SidecarTicket(%x) "+
+				"complete with state '%v'!", localTicket.ID[:],
+				fin.state)
 
 			// The ticket has been marked as finalized, so we'll
-			// update it as being completed in the database.
-			localTicket.State = sidecar.StateCompleted
+			// update it as being in a final state in the database.
+			localTicket.State = fin.state
 			if err := a.cfg.Driver.UpdateSidecar(localTicket); err != nil {
 				log.Errorf("unable to update ticket to "+
 					"complete state: %v", err)
 				return
 			}
 
-			// We'll also tear down the mailbox as well as we no
-			// longer need it anymore.
-			err := a.cfg.MailBox.DelAcctMailbox(
-				streamID, acct.TraderKey,
-			)
-			if err != nil {
-				log.Errorf("unable to reclaim mailbox: %v", err)
-				return
+			// Did we ever go into the registered state or further?
+			// If no recipient registered, then nobody would listen
+			// for the cancellation message.
+			switch {
+			// Our side canceled the ticket and a recipient
+			// registered for it. We need to inform them about the
+			// cancellation. They will then go ahead and remove the
+			// mailbox from their end.
+			case !fin.otherSide &&
+				fin.state == sidecar.StateCanceled &&
+				a.CurrentState() >= sidecar.StateRegistered:
+
+				// Sending a message doesn't block until it is
+				// received. To make sure we don't cancel the
+				// context right after we've sent the message
+				// (but maybe _before_ it is received), we use
+				// a background context here.
+				ctxb := context.Background()
+				err := a.cfg.MailBox.SendSidecarPkt(
+					ctxb, localTicket, false,
+				)
+				if err != nil {
+					log.Errorf("unable to send cancel "+
+						"msg to recipient: %v", err)
+					return
+				}
```

Comment:
I think we should fully tear down the goroutine here since at this point all negotiation has ceased. Same for the opposite role as well.